File: samples/multi-room-chat/app/chat/ChatEngine.scala
package chat

import akka.NotUsed
import akka.stream._
import akka.stream.scaladsl.{BroadcastHub, Flow, Keep, MergeHub, Sink, Source}
import play.api.libs.json.{Format, Json}
import play.engineio.EngineIOController
import play.api.libs.functional.syntax._
import play.socketio.SocketIO

import scala.collection.concurrent.TrieMap

object ChatProtocol {

  /**
    * A chat event, either a message, a join room, or a leave room event.
    */
  sealed trait ChatEvent {
    def user: Option[User]
    def room: String
  }

  case class ChatMessage(user: Option[User], room: String, message: String) extends ChatEvent
  object ChatMessage {
    implicit val format: Format[ChatMessage] = Json.format
  }

  case class JoinRoom(user: Option[User], room: String) extends ChatEvent
  object JoinRoom {
    implicit val format: Format[JoinRoom] = Json.format
  }

  case class LeaveRoom(user: Option[User], room: String) extends ChatEvent
  object LeaveRoom {
    implicit val format: Format[LeaveRoom] = Json.format
  }

  case class User(name: String)
  object User {
    // We're just encoding user as a simple string, not an object
    implicit val format: Format[User] = implicitly[Format[String]].inmap(User.apply, _.name)
  }

  import play.socketio.SocketIOEventCodec._

  val decoder = decoders(
    decodeJson[ChatMessage]("chat message"),
    decodeJson[JoinRoom]("join room"),
    decodeJson[LeaveRoom]("leave room")
  )

  val encoder = encoders {
    case _: ChatMessage => encodeJson[ChatMessage]("chat message")
    case _: JoinRoom => encodeJson[JoinRoom]("join room")
    case _: LeaveRoom => encodeJson[LeaveRoom]("leave room")
  }
}

class ChatEngine(socketIO: SocketIO)(implicit mat: Materializer) {

  import ChatProtocol._

  // All the chat rooms
  private val chatRooms = TrieMap.empty[String, Flow[ChatEvent, ChatEvent, _]]

  // This gets an existing chat room, or creates it if it doesn't exist
  private def getChatRoom(room: String) = {
    chatRooms.getOrElseUpdate(room, {
      // Each chat room is a merge hub merging messages into a broadcast hub.
      val (sink, source) = MergeHub.source[ChatEvent].toMat(BroadcastHub.sink[ChatEvent])(Keep.both).run
      Flow.fromSinkAndSource(sink, source)
    })
  }

  // Creates a chat flow for a user session
  def userChatFlow(user: User): Flow[ChatEvent, ChatEvent, NotUsed] = {

    // broadcast source and sink for demux/muxing multiple chat rooms in this one flow
    // They'll be provided later when we materialize the flow
    var broadcastSource: Source[ChatEvent, NotUsed] = null
    var mergeSink: Sink[ChatEvent, NotUsed] = null

    Flow[ChatEvent] collect Function.unlift {
      case JoinRoom(_, room) =>
        val roomFlow = getChatRoom(room)

        // Add the room to our flow
        broadcastSource
          // Ensure only messages for this room get there
          .filter(_.room == room)
          // Take until we get a leave room message.
          .takeWhile(!_.isInstanceOf[LeaveRoom])
          // We ensure that a join room message is sent first, and a leave room message is sent when we disconnect.
          // The reason we filter the previous LeaveRoom message out, and add one here, is so that when an unclean
          // disconnect happens (disconnect without sending leave room first), we always still send a leave room.
          .prepend(Source.single(JoinRoom(Some(user), room)))
          .concat(Source.single(LeaveRoom(Some(user), room)))
          // And send it through the room flow
          .via(roomFlow)
          // And now we want to cancel this end of the flow when we see the leave room message for us.
          .takeWhile({
            case LeaveRoom(Some(`user`), `room`) => false
            case _ => true
          }, inclusive = true)
          .runWith(mergeSink)

        // And ignore the original JoinRoom message, we've already ensured it gets prepended to our flow of messages
        // into the room
        None

      case LeaveRoom(_, room) =>
        Some(LeaveRoom(Some(user), room))

      case ChatMessage(_, room, message) =>
        Some(ChatMessage(Some(user), room, message))

    } via {
      Flow.fromSinkAndSourceCoupledMat(BroadcastHub.sink[ChatEvent], MergeHub.source[ChatEvent]) { (source, sink) =>
        broadcastSource = source
        mergeSink = sink
        NotUsed
      }
    }
  }

  val controller: EngineIOController = socketIO.builder
    .onConnect { (request, sid) =>
      // Extract the username from the header
      val username = request.getQueryString("user").getOrElse {
        throw new RuntimeException("No user parameter")
      }
      // And return the user, this will be the data for the session that we can read when we add a namespace
      User(username)
    }.addNamespace(decoder, encoder) {
      case (session, chat) if chat.split('?').head == "/chat" => userChatFlow(session.data)
    }
    .createController()
}

File: samples/java/multi-room-chat/app/chat/ChatEngine.java
/*
 * Copyright (C) 2017-2019 Lightbend Inc. <https://www.lightbend.com>
 */
package chat;

import akka.NotUsed;
import akka.japi.Pair;
import akka.stream.Materializer;
import akka.stream.javadsl.*;
import lombok.val;
import play.engineio.EngineIOController;
import play.socketio.javadsl.SocketIO;
import play.socketio.javadsl.SocketIOEventCodec;

import javax.inject.Inject;
import javax.inject.Provider;
import javax.inject.Singleton;

import chat.ChatEvent.*;

import java.util.Optional;
import java.util.concurrent.ConcurrentHashMap;

@Singleton
public class ChatEngine implements Provider<EngineIOController> {

  private final EngineIOController controller;
  private final Materializer materializer;
  // All the chat rooms
  private final ConcurrentHashMap<String, Pair<Sink<ChatEvent, NotUsed>, Source<ChatEvent, NotUsed>>> chatRooms =
      new ConcurrentHashMap<>();

  @Inject
  @SuppressWarnings("unchecked")
  public ChatEngine(SocketIO socketIO, Materializer materializer) {
    this.materializer = materializer;

    // Here we define our codec. We're serializing our events to/from json.
    val codec = new SocketIOEventCodec<ChatEvent, ChatEvent>() {
      {
        addDecoder("chat message", decodeJson(ChatMessage.class));
        addDecoder("join room", decodeJson(JoinRoom.class));
        addDecoder("leave room", decodeJson(LeaveRoom.class));

        addEncoder("chat message", ChatMessage.class, encodeJson());
        addEncoder("join room", JoinRoom.class, encodeJson());
        addEncoder("leave room", LeaveRoom.class, encodeJson());
      }
    };

    controller = socketIO.createBuilder()
        .onConnect((request, sid) -> {
          // Extract the username from the header
          val username = request.getQueryString("user");
          if (username == null) {
            throw new RuntimeException("No user parameter");
          }
          // And return the user, this will be the data for the session that we can read when we add a namespace
          return new User(username);

        }).addNamespace(codec, (session, chat) -> {
          if (chat.split("\\?")[0].equals("/chat")) {
            return Optional.of(createFlow(session.data()));
          } else {
            return Optional.empty();
          }
        })
        .createController();
  }

  // This gets an existing chat room, or creates it if it doesn't exist
  private Flow<ChatEvent, ChatEvent, NotUsed> getChatRoom(String room, User user) {

    val pair = chatRooms.computeIfAbsent(room, (r) -> {
      // Each chat room is a merge hub merging messages into a broadcast hub.
      return MergeHub.of(ChatEvent.class).toMat(BroadcastHub.of(ChatEvent.class), Keep.both()).run(materializer);
    });

    // A coupled sink and source ensures if either side is cancelled/completed, the other will be too.
    return Flow.fromSinkAndSourceCoupled(
        Flow.<ChatEvent>create()
            // Add the join and leave room events
            .concat(Source.<ChatEvent>single(new LeaveRoom(user, room)))
            .prepend(Source.<ChatEvent>single(new JoinRoom(user, room)))
            .to(pair.first()),
        pair.second()
    );
  }

  private Flow<ChatEvent, ChatEvent, NotUsed> createFlow(User user) {
    // broadcast source and sink for demux/muxing multiple chat rooms in this one flow
    // They'll be provided later when we materialize the flow
    Source<ChatEvent, NotUsed>[] broadcastSource = new Source[1];
    Sink<ChatEvent, NotUsed>[] mergeSink = new Sink[1];

    // Create a chat flow for a user session
    return Flow.<ChatEvent>create().map(event -> {
      if (event instanceof JoinRoom) {
        val room = event.getRoom();
        val roomFlow = getChatRoom(room, user);

        // Add the room to our flow
        broadcastSource[0]
            // Ensure only messages for this room get there.
            // Also filter out JoinRoom messages, since there's a race condition as to whether it will
            // actually get here or not, so we explicitly add it below.
            .filter(e -> e.getRoom().equals(room) && !(e instanceof JoinRoom))
            // Take until we get a leave room message.
            .takeWhile(e -> !(e instanceof LeaveRoom))
            // We ensure that a join room message is sent first
            // And send it through the room flow
            .via(roomFlow)
            // Re-add the leave room here, since it was filtered out before
            .concat(Source.<ChatEvent>single(new LeaveRoom(user, room)))
            // And run it with the merge sink
            .runWith(mergeSink[0], materializer);
        return event;
      } else if (event instanceof ChatMessage) {
        // Add the user
        return new ChatMessage(user, event.getRoom(), ((ChatMessage) event).getMessage());
      } else {
        return event;
      }
    }).via(
        Flow.fromSinkAndSourceCoupledMat(BroadcastHub.of(ChatEvent.class), MergeHub.of(ChatEvent.class), (source, sink) -> {
          broadcastSource[0] = source;
          mergeSink[0] = sink;
          return NotUsed.getInstance();
        })
    );
  }

  public EngineIOController get() {
    return controller;
  }
}